Objects/abstract.c

```diff
@@ -199,6 +199,30 @@ PyObject_GetItem(PyObject *o, PyObject *key)
     return type_error("'%.200s' object is not subscriptable", o);
 }
 
+int
+PyMapping_GetOptionalItem(PyObject *o, PyObject *key, PyObject **pvalue)
+{
+    if (PyDict_CheckExact(o)) {
+        *pvalue = PyDict_GetItemWithError(o, key);  /* borrowed */
+        if (*pvalue) {
+            Py_INCREF(*pvalue);
+            return 1;
+        }
+        return PyErr_Occurred() ? -1 : 0;
```

Comment:
PyDict_GetItemRef() (PR #106005) would avoid having to call PyErr_Occurred() here :-)

Reply:
Yes, it would be simply:
```c
return PyDict_GetItemRef(obj, key, result);
```
And correspondingly the use of `PyDict_GetItemRef()` instead of `PyDict_GetItemWithError()` can simplify a lot of other code.